core/src/main/scala/org/apache/spark/sql/rapids/tool/store/StageModelManager.scala

```diff
@@ -0,0 +1,164 @@
+    val sModel = StageModel(stageInfo, currentAttempts.get(stageInfo.attemptNumber()))
+    currentAttempts.put(stageInfo.attemptNumber(), sModel)
+    sModel
+  }
+
+  // Internal method to update the accumulatorMap using the current stage model.
+  private def addAccumIdsToStage(stageModel: StageModel): Unit = {
+    val sInfoAccumIds = stageModel.getSInfoAccumIds
+    if (sInfoAccumIds.nonEmpty) {
+      sInfoAccumIds.foreach { accumId =>
+        val stageIds = accumIdToStageId.getOrElseUpdate(accumId, TreeSet[Int]())
+        accumIdToStageId.put(accumId, stageIds + stageModel.sId)
+      }
+    }
+  }
+
+  // Used to retrieve a stage model and does not create a new instance if it does not exist.
+  def getStage(stageId: Int, attemptId: Int): Option[StageModel] = {
+    stageIdToInfo.get(stageId).flatMap(_.get(attemptId))
+  }
+
+  // Used to retrieve list of stages by stageId (in case multiple attempts exist)
+  def getStagesByIds(stageIds: Iterable[Int]): Iterable[StageModel] = {
+    stageIds.flatMap { stageId =>
+      stageIdToInfo.get(stageId).map(_.values)
+    }.flatten
+  }
+
+  // Returns all stages that have failed
+  def getFailedStages: Iterable[StageModel] = {
+    getAllStages.filter(_.hasFailed)
+  }
+
+  // Shortcut to get the duration of a stage by stageId
+  @WallClock
+  @Calculated("Sum all the WallaClockDuration for the given stageId")
+  def getDurationById(stageId: Int): Long = {
+    stageIdToInfo.get(stageId).map { attempts =>
+      attempts.values.map(_.getDuration).sum
+    }.getOrElse(0L)
+  }
+
+  // Remove all Stages by stageId
+  def removeStages(stageIds: Iterable[Int]): Unit = {
+    stageIdToInfo --= stageIds
+  }
+
+  /**
+   * Given a Spark.StageInfo, this method will return an existing StageModel that has
+   * (stageId, attemptId) if it exists. Otherwise, it will create a new instance of StageModel.
+   * Once the stageModel instance is obtained, it will update the accumulator mapping based on the
+   * elements of sInfo.accumulables.
+   *
+   * @param sInfo a snapshot from StageInfo captured from StageSubmitted/StageCompleted-events
+   * @return existing or new instance of StageModel with (sInfo.stageId, sInfo.attemptID)
+   */
+  def addStageInfo(sInfo: StageInfo): StageModel = {
+    val stage = getOrCreateStage(sInfo)
+    addAccumIdsToStage(stage)
+    stage
+  }
+
+  /**
+   * Returns a mapping between AccumulatorID and a single stageId (1-to-1) by taking the head of
+   * the list.
+   * That getter is used as a temporary hack to avoid callers that expect a 1-to-1 mapping between
+   * accumulators and stages. i.e., GenerateDot.writeDotGraph expects a 1-to-1 mapping but it is
+   * rarely used for now.
+   *
+   * @return a Map of AccumulatorID to StageId
+   */
+  def getAccumToSingleStage(): Map[Long, Int] = {
+    accumIdToStageId.map { case (accumId, stageIds) =>
+      accumId -> stageIds.head
+    }.toMap
```

Comment:
Nit: We could use collect here, also check if head has some value. Since this function is rarely used, I don't think this would have caused any issue:

```
    accumIdToStageId.collect {
      case (accumId, stageIds) if stageIds.nonEmpty => accumId -> stageIds.head
    }
```